File: FastMines_core/UnitTest.FastMines.Core/Core/MosaicHelperTest.cs
using System;
using System.Threading.Tasks;
using NUnit.Framework;
using fmg.core.types;
using fmg.common.geom;
using fmg.common;
using fmg.common.ui;

namespace fmg.core.mosaic {

    public class MosaicHelperTest {

        [SetUp]
        public void Setup() {
            LoggerSimple.Put(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>");
            LoggerSimple.Put("> MosaicHelperTest::Setup");

            MosaicModelTest.StaticInitializer();
        }

        [Test]
        public void FindSizeByArea_eMosaicSquare1_Test() {
            {
                var sizeClient = new SizeDouble(100, 100);
                var sizeField = MosaicHelper.FindSizeByArea(EMosaic.eMosaicSquare1, 100, sizeClient);

                Assert.AreEqual(new Matrisize(10, 10), sizeField);
            }
            {
                var sizeClient = new SizeDouble(500, 500);
                var sizeField = MosaicHelper.FindSizeByArea(EMosaic.eMosaicSquare1, 100, sizeClient);

                Assert.AreEqual(new Matrisize(50, 50), sizeField);
            }
            {
                var sizeClient = new SizeDouble(300, 700);
                var sizeField = MosaicHelper.FindSizeByArea(EMosaic.eMosaicSquare1, 100, sizeClient);

                Assert.AreEqual(new Matrisize(30, 70), sizeField);
            }
        }

        [Test]
        public void FindAreaBySize_eMosaicSquare1_Test1() {
            var sizeClientIn = new SizeDouble(100, 100);
            var sizeClientOut = sizeClientIn;
            var area = MosaicHelper.FindAreaBySize(EMosaic.eMosaicSquare1, new Matrisize(10, 10), ref sizeClientOut);

            // Assert.AreEqual(100, area);
            Assert.GreaterOrEqual(MosaicHelper.AreaPrecision, 100 - area);

            // Assert.AreEqual(sizeClientIn, sizeClientOut);
            Assert.GreaterOrEqual(MosaicHelper.AreaPrecision, sizeClientIn.Width - sizeClientOut.Width);
            Assert.GreaterOrEqual(MosaicHelper.AreaPrecision, sizeClientIn.Height - sizeClientOut.Height);
        }

        [Test]
        public void FindAreaBySize_eMosaicSquare1_Test2() {
            var sizeClientIn = new SizeDouble(200, 200);
            var sizeClientOut = sizeClientIn;
            var area = MosaicHelper.FindAreaBySize(EMosaic.eMosaicSquare1, new Matrisize(10, 20), ref sizeClientOut);

            // Assert.AreEqual(100, area);
            Assert.GreaterOrEqual(MosaicHelper.AreaPrecision, 100 - area);

            // Assert.AreEqual(new SizeDouble(100, 200), sizeClientOut);
            Assert.GreaterOrEqual(MosaicHelper.AreaPrecision, 100 - sizeClientOut.Width);
            Assert.GreaterOrEqual(MosaicHelper.AreaPrecision, sizeClientIn.Height - sizeClientOut.Height);
        }

        [Test]
        public void FindAreaBySize_Random_Test() {
            var r = ThreadLocalRandom.Current;

            for (var i = 0; i < 1000; ++i) {
                var sizeClientIn = new SizeDouble(100 + r.Next(500), 100 + r.Next(500));
                var sizeClientOut = sizeClientIn;
                var mosaicType = EMosaicEx.FromOrdinal(r.Next(EMosaicEx.GetValues().Length));
                var mSize = new Matrisize(5 + r.Next(30), 5 + r.Next(30));
                var area = MosaicHelper.FindAreaBySize(mosaicType, mSize, ref sizeClientOut);

                // Assert.AreEqual(100, area);
                //Assert.GreaterOrEqual(MosaicHelper.AreaPrecision, ??? - area);
                Assert.Less(0, area);

                var magicNumber = 8;
                Assert.IsTrue((MosaicHelper.AreaPrecision >= (sizeClientIn.Width  - sizeClientOut.Width ) / magicNumber) ||
                              (MosaicHelper.AreaPrecision >= (sizeClientIn.Height - sizeClientOut.Height) / magicNumber));
            }
        }


        [Test]
        public async Task FindAreaBySize_eMosaicTrapezoid3_Test() {
            var signal = new Signal();
            double area = -1;
            Factory.DEFERR_INVOKER(() => {
                var sizeClientIn = new SizeDouble(169.90442448471225, 313.90196868082262);
                SizeDouble sizeClientOut = sizeClientIn;
                area = MosaicHelper.FindAreaBySize(EMosaic.eMosaicTrapezoid3, new Matrisize(4, 2), ref sizeClientOut);

                signal.Set();
            });

            Assert.IsTrue(await signal.Wait(TimeSpan.FromSeconds(1)));
            Assert.IsTrue(area > 0);
        }

    }

}

File: FastMines_core/src_test/fmg/core/mosaic/MosaicHelperTest.java
package fmg.core.mosaic;

import java.util.concurrent.ThreadLocalRandom;

import org.junit.*;

import fmg.common.Logger;
import fmg.common.geom.Matrisize;
import fmg.common.geom.SizeDouble;
import fmg.core.types.EMosaic;

public class MosaicHelperTest {

    @BeforeClass
    public static void setup() {
        Logger.info(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>");
        Logger.info("> MosaicHelperTest::setup");
    }

    @Before
    public void before() {
        Logger.info("======================================================");
    }
    @AfterClass
    public static void after() {
        Logger.info("======================================================");
        Logger.info("< MosaicHelperTest closed");
        Logger.info("<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<");
    }


    @Test
    public void findSizeByAreaMosaicSquare1Test() {
        Logger.info("> MosaicHelperTest::findSizeByAreaMosaicSquare1Test");

        {
            SizeDouble sizeClient = new SizeDouble(100, 100);
            Matrisize sizeField = MosaicHelper.findSizeByArea(EMosaic.eMosaicSquare1, 100, sizeClient);

            Assert.assertEquals(new Matrisize(10, 10), sizeField);
        }
        {
            SizeDouble sizeClient = new SizeDouble(500, 500);
            Matrisize sizeField = MosaicHelper.findSizeByArea(EMosaic.eMosaicSquare1, 100, sizeClient);

            Assert.assertEquals(new Matrisize(50, 50), sizeField);
        }
        {
            SizeDouble sizeClient = new SizeDouble(300, 700);
            Matrisize sizeField = MosaicHelper.findSizeByArea(EMosaic.eMosaicSquare1, 100, sizeClient);

            Assert.assertEquals(new Matrisize(30, 70), sizeField);
        }
    }

    @Test
    public void findAreaBySizeMosaicSquare1Test1() {
        Logger.info("> MosaicHelperTest::findAreaBySizeMosaicSquare1Test1");

        SizeDouble sizeClientIn = new SizeDouble(100, 100);
        SizeDouble sizeClientOut = new SizeDouble();
        double area = MosaicHelper.findAreaBySize(EMosaic.eMosaicSquare1, new Matrisize(10, 10), sizeClientIn, sizeClientOut);

        // Assert.assertEquals(100, area);
        Assert.assertTrue(MosaicHelper.AREA_PRECISION >= 100 - area);

        // Assert.assertEquals(sizeClientIn, sizeClientOut);
        Assert.assertTrue(MosaicHelper.AREA_PRECISION >= sizeClientIn.width  - sizeClientOut.width);
        Assert.assertTrue(MosaicHelper.AREA_PRECISION >= sizeClientIn.height - sizeClientOut.height);
    }

    @Test
    public void findAreaBySizeMosaicSquare1Test2() {
        Logger.info("> MosaicHelperTest::findAreaBySizeMosaicSquare1Test2");

        SizeDouble sizeClientIn = new SizeDouble(200, 200);
        SizeDouble sizeClientOut = new SizeDouble();
        double area = MosaicHelper.findAreaBySize(EMosaic.eMosaicSquare1, new Matrisize(10, 20), sizeClientIn, sizeClientOut);

        // Assert.assertEquals(100, area);
        Assert.assertTrue(MosaicHelper.AREA_PRECISION >= 100 - area);

        // Assert.assertEquals(new SizeDouble(100, 200), sizeClientOut);
        Assert.assertTrue(MosaicHelper.AREA_PRECISION >= 100 - sizeClientOut.width);
        Assert.assertTrue(MosaicHelper.AREA_PRECISION >= sizeClientIn.height - sizeClientOut.height);
    }

    @Test
    public void findAreaBySizeMosaicSquare1Test3() {
        Logger.info("> MosaicHelperTest::findAreaBySizeMosaicSquare1Test3");

        {
            SizeDouble sizeClientIn = new SizeDouble(200, 400);
            SizeDouble sizeClientOut = new SizeDouble();
            double area = MosaicHelper.findAreaBySize(EMosaic.eMosaicSquare1, new Matrisize(10, 10), sizeClientIn, sizeClientOut);

            // Assert.assertEquals(100, area);
            Assert.assertTrue(MosaicHelper.AREA_PRECISION >= 400 - area);

            // Assert.assertEquals(new SizeDouble(100, 200), sizeClientOut);
            Assert.assertTrue(MosaicHelper.AREA_PRECISION >= 200 - sizeClientOut.width);
            Assert.assertTrue(MosaicHelper.AREA_PRECISION >= sizeClientIn.width - sizeClientOut.width);
        }
        {
            SizeDouble sizeClientIn = new SizeDouble(400, 200);
            SizeDouble sizeClientOut = new SizeDouble();
            double area = MosaicHelper.findAreaBySize(EMosaic.eMosaicSquare1, new Matrisize(10, 10), sizeClientIn, sizeClientOut);

            // Assert.assertEquals(100, area);
            Assert.assertTrue(MosaicHelper.AREA_PRECISION >= 400 - area);

            // Assert.assertEquals(new SizeDouble(100, 200), sizeClientOut);
            Assert.assertTrue(MosaicHelper.AREA_PRECISION >= 200 - sizeClientOut.height);
            Assert.assertTrue(MosaicHelper.AREA_PRECISION >= sizeClientIn.height - sizeClientOut.height);
        }
    }

    @Test
    public void findAreaBySizeRandomTest() {
        Logger.info("> MosaicHelperTest::findAreaBySizeRandomTest");

        ThreadLocalRandom r = ThreadLocalRandom.current();

        for (int i = 0; i < 1000; ++i) {
            SizeDouble sizeClientIn = new SizeDouble(100 + r.nextInt(500), 100 + r.nextInt(500));
            SizeDouble sizeClientOut = new SizeDouble();
            EMosaic mosaicType = EMosaic.fromOrdinal(r.nextInt(EMosaic.values().length));
            Matrisize mSize = new Matrisize(5 + r.nextInt(30), 5 + r.nextInt(30));
            double area = MosaicHelper.findAreaBySize(mosaicType, mSize, sizeClientIn, sizeClientOut);

            //Assert.assertEquals(100, area);
            //Assert.assertTrue(MosaicHelper.AreaPrecision >= (??? - area));
            Assert.assertTrue(area > 0);

            int magicNumber = 8;
            try {
                Assert.assertTrue((MosaicHelper.AREA_PRECISION >= (sizeClientIn.width  - sizeClientOut.width )/magicNumber) ||
                                  (MosaicHelper.AREA_PRECISION >= (sizeClientIn.height - sizeClientOut.height)/magicNumber));
            } catch (Throwable ex) {
                Logger.error("Failed for sizeClientIn=" + sizeClientIn
                                   + "; sizeClientOut=" + sizeClientOut
                                   + "; mosaicType=" + mosaicType
                                   + "; mSize=" + mSize
                                   + "; area=" + area);
                throw ex;
            }
        }
    }

    @Test
    public void findAreaBySizeMosaicTrapezoid3Test() {
        Logger.info("> MosaicHelperTest::findAreaBySizeMosaicTrapezoid3Test");

        SizeDouble sizeClientIn = new SizeDouble(169.90442448471225, 313.90196868082262);
        SizeDouble sizeClientOut = new SizeDouble();
        var area = MosaicHelper.findAreaBySize(EMosaic.eMosaicTrapezoid3, new Matrisize(4, 2), sizeClientIn, sizeClientOut);

        Assert.assertTrue(area > 0);
    }

    @Test
    public void findAreaBySizeMosaicTriangle1Test() {
        Logger.info("> MosaicHelperTest::findAreaBySizeMosaicTriangle1Test");

        SizeDouble sizeClientIn = new SizeDouble(186.89486693318347, 294.28309563827116);
        SizeDouble sizeClientOut = new SizeDouble();
        double area= MosaicHelper.findAreaBySize(EMosaic.eMosaicTriangle1, new Matrisize(3, 2), sizeClientIn, sizeClientOut);
        Assert.assertTrue(area > 0);
    }

}
